core/shared/src/main/scala/org/http4s/Service.scala

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2013 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s
+
+import cats.Alternative
+import cats.Applicative
+import cats.Monad
+import cats.StackSafeMonad
+import cats.effect.kernel.Resource
+import cats.mtl.Local
+
+import Service._
+
+sealed abstract class Service[F[_], -A, B] {
+  protected[Service] def run: A => Resource[F, Option[B]]
+
+  def runOrElse(b: B): A => Resource[F, B] =
+    a => run(a).map(_.getOrElse(b))
+
+  def map[C](f: B => C): Service[F, A, C] =
+    apply((a: A) => run(a).map(_.map(f)))
+
+  def flatMap[AA <: A, C](f: B => Service[F, AA, C]) =
+    apply((a: AA) =>
+      run(a).flatMap(_ match {
+        case Some(b) => f(b).run(a)
+        case None => Resource.pure[F, Option[C]](None)
+      })
+    )
+
+  def orElse[AA <: A](default: => Service[F, AA, B]): Service[F, AA, B] =
+    apply(a =>
+      run(a).flatMap {
+        case some: Some[_] => Resource.pure(some)
+        case None => default.run(a)
+      }
+    )
+
+  def local[C](f: C => A): Service[F, C, B] =
+    apply(c => run(f(c)))
+}
+
+object Service extends ServiceInstances {
+  private[Service] final case class Run[F[_], -A, B](run: A => Resource[F, Option[B]])
+      extends Service[F, A, B]
+
+  def apply[F[_], A, B](run: A => Resource[F, Option[B]]): Run[F, A, B] =
+    Run(run)
+
+  def pure[F[_], A, B](b: B): Service[F, A, B] =
+    apply(Function.const(Resource.pure(Some(b))))
+
+  def pass[F[_], A, B]: Service[F, A, B] =
+    apply(Function.const(Resource.pure(None)))
```

Comment:
We can cache this and cast 😅

Reply:
Yes, if the applyOrElse encoding falls on its face.